nixos/shadow: refactor login.defs config options

GeorgesAlkhouri opened the pull request:
###### Description of changes

Add a `security.loginDefs.settings` option for the login.defs file to provide enhanced configuration abilities.

E.g. add hardening options to login.defs file
```
security.loginDefs.settings.ENCRYPT_METHOD = "SHA512";
security.loginDefs.settings.SHA_CRYPT_MIN_ROUNDS = 999999999;
security.loginDefs.settings.SHA_CRYPT_MAX_ROUNDS =  999999999;
security.loginDefs.settings.PASS_MAX_DAYS = 365;
security.loginDefs.settings.PASS_MIN_DAYS = 1;
```

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

JamieMagee commented:
`ENCRYPT_METHOD` should to be kept in sync with pam:

https://github.com/shadow-maint/shadow/blob/1d936c968a213319eff46e78284c9dd45d6c79dd/etc/login.defs#L334-L335

See:

https://github.com/NixOS/nixpkgs/blob/b3a1a8050f59bd94dc00685593014b2108cb842d/nixos/modules/security/pam.nix#L608

JamieMagee commented:
Result of `nixpkgs-review pr 208363` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tests.testers.nixosTest-example</li>
  </ul>
</details>

GeorgesAlkhouri commented:
> `ENCRYPT_METHOD` should to be kept in sync with pam:
> 
> 
> 
> https://github.com/shadow-maint/shadow/blob/1d936c968a213319eff46e78284c9dd45d6c79dd/etc/login.defs#L334-L335
> 
> 
> 
> See:
> 
> 
> 
> https://github.com/NixOS/nixpkgs/blob/b3a1a8050f59bd94dc00685593014b2108cb842d/nixos/modules/security/pam.nix#L608

Yes you're right. But the idea of the extraConfigs attribute was to provide a quick way to add options to login.defs where the user is responsible what to put in there.

Also the hash algorithm for passwords in pam.nix is set implicitly and there is no option for it in the module.

But sure I could provide an option for `ENCRYPT_METHOD` if used, mirrors the value in pam.nix, hard coded of course.

GeorgesAlkhouri commented:
Something else todo here or do we just have to wait for a second review?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1876

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -83,9 +94,14 @@ in
     environment.etc =
-      { # /etc/login.defs: global configuration for pwdutils.  You
+      {
+        # /etc/login.defs: global configuration for pwdutils.  You
         # cannot login without it!
```
```suggestion
        # /etc/login.defs: global configuration for pwdutils.
        # You cannot login without it!
```

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -58,6 +58,17 @@ in
 
```
While we are at it: Why not resolve the todo in line 42?

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -58,6 +58,17 @@ in
 
+    security.loginDefs.extraConfig = mkOption {
+      default = "";
+      type = types.lines;
```
Why not use a rfc42 style type here?

GeorgesAlkhouri replied:
And therefore also melt the hard coded [login.defs](https://github.com/NixOS/nixpkgs/blob/9f1c3c7dad56a197a55def1963d0737ae74f0b7b/nixos/modules/programs/shadow.nix#L19-L44) into an ini settings format with `pkgs.formats.ini`? This would also tackle your other suggestion with the `CHFN_RESTRICT` option.

GeorgesAlkhouri replied:
ok, not ini I wrote a custom toKeyValue generator to reflect the login.defs config spec.

GeorgesAlkhouri commented:
ok, I added settings option and reflected the existing default options.  Also I reformatted the source code file. 
Todo would be to add better types reflecting the possible options, e.g. for `security.loginDefs.chfnRestrict`. @SuperSandro2000 , @JamieMagee what do you say.

SuperSandro2000 commented:
Please remove the merge commit from the PR. it makes rebasing a nighmare when the merge conflict falls onto it.

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
-
 with lib;
-
+with types;
```
Please make this more local.

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
+    mkKeyValue = generators.mkKeyValueDefault { } " ";
+  };
+  configFile = pkgs.writeText "login.defs" (toKeyValue cfg.settings);
+  yesNoType = enum [ "yes" "no" ];
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
+              description = mdDoc
+                "Indicate if login is allowed if we can't cd to the home directory.";
+              default = "yes";
+              type = yesNoType;
```
```suggestion
              type = enum [ "yes" "no" ];;
```

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
+  toKeyValue = generators.toKeyValue {
+    mkKeyValue = generators.mkKeyValueDefault { } " ";
+  };
+  configFile = pkgs.writeText "login.defs" (toKeyValue cfg.settings);
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
-        groupdel = { rootOK = true; };
-        login = { startSession = true; allowNullPassword = true; showMotd = true; updateWtmp = true; };
-        chpasswd = { rootOK = true; };
+      "login.defs".source = configFile;
```
```suggestion
      "login.defs".source = pkgs.writeText "login.defs" (toKeyValue cfg.settings);;
```

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
+          See login.defs(5) man page for available options.
+        '';
+        type = submodule {
+          freeformType = settingsFormat.type;
```
```suggestion
          freeformType = (pkgs.formats.keyValue { }).type;
```

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
+          */
+          options = {
+            DEFAULT_HOME = mkOption {
+              description = mdDoc
+                "Indicate if login is allowed if we can't cd to the home directory.";
```
```suggestion
              description = mdDoc "Indicate if login is allowed if we can't cd to the home directory.";
```
same for the other options

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,128 +1,220 @@
+            GID_MAX = mkOption {
+              description = mdDoc
+                "Range of group IDs used for the creation of regular groups by useradd, groupadd, or newusers.";
+              default = 29999;
```
We should add some asserts if min and max make sense

GeorgesAlkhouri committed:
nixos/shadow: refactor login.defs config options

Add a settings option for the login.defs file
to provide enhanced configuration abilities.

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,129 +1,210 @@
+        };
+      in
+      {
+        su = mkSetuidRoot "${pkgs.shadow.su}/bin/su";
+        sg = mkSetuidRoot "${pkgs.shadow.out}/bin/sg";
+        newgrp = mkSetuidRoot "${pkgs.shadow.out}/bin/newgrp";
+        newuidmap = mkSetuidRoot "${pkgs.shadow.out}/bin/newuidmap";
+        newgidmap = mkSetuidRoot "${pkgs.shadow.out}/bin/newgidmap";
+      } // optionalAttrs config.users.mutableUsers {
+        chsh = mkSetuidRoot "${pkgs.shadow.out}/bin/chsh";
+        passwd = mkSetuidRoot "${pkgs.shadow.out}/bin/passwd";
```
Should we add a package option?

GeorgesAlkhouri replied:
Could you elaborate more?

SuperSandro2000 replied:
see https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/nexttrace.nix#L11

SuperSandro2000 commented on nixos/modules/programs/shadow.nix:
```diff
@@ -1,129 +1,210 @@
+      # Note: useradd, groupadd etc. aren't setuid root, so it
+      # doesn't really matter what the PAM config says as long as it
+      # lets root in.
+      useradd = { rootOK = true; };
+      usermod = { rootOK = true; };
+      userdel = { rootOK = true; };
+      groupadd = { rootOK = true; };
+      groupmod = { rootOK = true; };
+      groupmems = { rootOK = true; };
+      groupdel = { rootOK = true; };
```
```suggestion
      useradd.rootOK = true;
      usermod.rootOK = true;
      userdel.rootOK = true;
      groupadd.rootOK = true;
      groupmod.rootOK = true;
      groupmems.rootOK = true;
      groupdel.rootOK = true;
```
we can cleanup this a little

GeorgesAlkhouri commented:
yes sure.

aanderse reviewed (approved):
This is great! Thanks for contributing. Looking forward to when we can merge this soon.

GeorgesAlkhouri commented:
As the release is through, can we merge?

aanderse commented:
Thanks for the ping. I didn't test this and only looked it over from a high level point of view... is everything solid, tested, and good to merge? If so, let's do it!

GeorgesAlkhouri commented:
Yes, I have tested this on my machine, the values for the login.defs config are kept the same as before. So the behavior should only change if the user changes the values.

I could add this NixOS test though, which only checks if the values are set correctly and the system is starting. What do you think?
```nix
import ./make-test-python.nix ({ pkgs, ... }:

{
  name = "login.defs";
  meta = with pkgs.lib.maintainers; {
    maintainers = [ georgesalkhouri ];
  };

  nodes.machine =
    { pkgs, ... }:
    {


      security.loginDefs.settings = {
        DEFAULT_HOME = "yes";
        ENCRYPT_METHOD = "SHA512";
        SYS_UID_MIN = 400;
        SYS_UID_MAX = 999;
        UID_MIN = 1000;
        UID_MAX = 29999;
        SYS_GID_MIN = 400;
        SYS_GID_MAX = 999;
        GID_MIN = 1000;
        GID_MAX = 29999;
        TTYGROUP = "tty";
        TTYPERM = "0620";
        UMASK = "077";
      };
      security.loginDefs.chfnRestrict = "frwh";

    };

  testScript = ''

      machine.wait_for_unit("multi-user.target")
      machine.wait_until_succeeds("pgrep -f 'agetty.*tty1'")

      with subtest("Check if security.loginDefs options are set correctly"):
          machine.succeed("grep -q 'DEFAULT_HOME yes' /etc/login.defs")
          machine.succeed("grep -q 'ENCRYPT_METHOD SHA512' /etc/login.defs")
          machine.succeed("grep -q 'SYS_UID_MIN 400' /etc/login.defs")
          machine.succeed("grep -q 'SYS_UID_MAX 999' /etc/login.defs")
          machine.succeed("grep -q 'UID_MIN 1000' /etc/login.defs")
          machine.succeed("grep -q 'UID_MAX 29999' /etc/login.defs")
          machine.succeed("grep -q 'SYS_GID_MIN 400' /etc/login.defs")
          machine.succeed("grep -q 'SYS_GID_MAX 999' /etc/login.defs")
          machine.succeed("grep -q 'GID_MIN 1000' /etc/login.defs")
          machine.succeed("grep -q 'GID_MAX 29999' /etc/login.defs")
          machine.succeed("grep -q 'TTYGROUP tty' /etc/login.defs")
          machine.succeed("grep -q 'TTYPERM 0620' /etc/login.defs")
          machine.succeed("grep -q 'UMASK 077' /etc/login.defs")
          machine.succeed("grep -q 'CHFN_RESTRICT frwh' /etc/login.defs")
  '';
})

```

GeorgesAlkhouri commented:
@aanderse, so I think the additional test is not needed and this PR can be merged.

aanderse commented:
Great! Thanks for contributing, and sorry for the delay.
